iohub/dataset_base.py

```diff
@@ -0,0 +1,176 @@
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Mapping
+from pathlib import Path
+from types import TracebackType
+from typing import Any, Iterable, Optional, Type, Union
+
+import numpy as np
+from numpy.typing import ArrayLike
+
+_AXES_PREFIX = ["T", "C", "Z", "Y", "X"]
+
+
+class BaseFOV(ABC):
+    _root: Path
+
+    @property
+    @abstractmethod
+    def axes_names(self) -> list[str]:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def channels(self) -> list[str]:
+        raise NotImplementedError
+
+    def channel_index(self, key: str) -> int:
+        """Return index of given channel."""
+        return self.channels.index(key)
+
+    def _missing_axes(self) -> list[int]:
+        """Return sorted indices of missing axes."""
+        if len(self.axes_names) == 5:
+            return []
+
+        elif len(self.axes_names) > 5:
+            raise ValueError(
+                f"{self.__name__} does not support more than 5 axes. "
+                f"Found {len(self.axes_names)}"
+            )
+
+        axes = set(ax[:1].upper() for ax in self.axes_names)
+
+        missing = []
+        for i, ax in enumerate(_AXES_PREFIX):
+            if ax not in axes:
+                missing.append(i)
+
+        return missing
+
+    def _pad_missing_axes(
+        self,
+        seq: Union[list[Any], tuple[Any]],
+        value: Any,
+    ) -> Union[list[Any], tuple[Any]]:
+        """Pads ``seq`` with ``value`` in the missing axes positions."""
+
+        if isinstance(seq, tuple):
+            is_tuple = True
+            seq = list(seq)
+        else:
+            is_tuple = False
+
+        for i in self._missing_axes():
+            seq.insert(i, value)
+
+        if is_tuple:
+            seq = tuple(seq)
+
+        assert len(seq) == len(_AXES_PREFIX)
+
+        return seq
+
+    @property
+    @abstractmethod
+    def shape(self) -> tuple[int, int, int, int, int]:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(
+        self,
+        key: Union[int, slice, tuple[Union[int, slice], ...]],
+    ) -> ArrayLike:
+        """
+        Returned object must support the ``__array__`` interface,
+        so that ``np.asarray(...)`` will work.
+        """
+        raise NotImplementedError
+
+    @property
+    def ndim(self) -> int:
+        return 5
+
+    @property
+    @abstractmethod
+    def dtype(self) -> np.dtype:
+        raise NotImplementedError
+
+    @property
```

Comment:
`BaseFOV` has `zyx_scale` and `channels/channel_names`, but no property related to the time dimension. 

Maybe `def t_scale(self) -> float`? Or maybe we want to support irregular time steps with `def times(self) -> list[float]`?

Reply:
Do you this the `list[float]` should be enforced by the API, or is it very specific to some datasets?

Reply:
@ziw-liu @talonchandler What do you think about ditching `zyx_scale` and having `scale` that returns a 5-tuple of floats?

Reply:
@ziw-liu , I see. I was expecting `Position.scale` to return the same, but it might not be 5-dimensional, right? This might lead to confusion in the future.